core/solver/multigrid.cpp

```diff
@@ -581,61 +531,126 @@ void Multigrid::generate()
+            },
+            b, x);
+    };
+    auto first_mg_level = this->get_mg_level_list().front();
+    run<gko::multigrid::EnableMultigridLevel, float, double,
+        std::complex<float>, std::complex<double>>(first_mg_level, lambda, b,
+                                                   x);
+}
+
+
+void Multigrid::apply_impl(const LinOp* alpha, const LinOp* b,
+                           const LinOp* beta, LinOp* x) const
+{
+    this->apply_with_initial_guess(alpha, b, beta, x,
+                                   this->get_default_initial_guess());
+}
+
+
+void Multigrid::apply_with_initial_guess_impl(const LinOp* alpha,
+                                              const LinOp* b, const LinOp* beta,
+                                              LinOp* x,
+                                              initial_guess_mode guess) const
+{
+    if (!this->get_system_matrix()) {
+        return;
+    }
+
+    auto lambda = [this, guess](auto mg_level, auto alpha, auto b, auto beta,
+                                auto x) {
+        using value_type = typename std::decay_t<
+            gko::detail::pointee<decltype(mg_level)>>::value_type;
+        experimental::precision_dispatch_real_complex_distributed<value_type>(
+            [this, guess](auto dense_alpha, auto dense_b, auto dense_beta,
+                          auto dense_x) {
+                prepare_initial_guess(dense_b, dense_x, guess);
+                auto x_clone = dense_x->clone();
+                this->apply_dense_impl(dense_b, x_clone.get(), guess);
+                dense_x->scale(dense_beta);
+                dense_x->add_scaled(dense_alpha, x_clone.get());
+            },
+            alpha, b, beta, x);
+    };
+    auto first_mg_level = this->get_mg_level_list().front();
+    run<gko::multigrid::EnableMultigridLevel, float, double,
+        std::complex<float>, std::complex<double>>(first_mg_level, lambda,
+                                                   alpha, b, beta, x);
+}
+
+
+template <typename VectorType>
+void Multigrid::apply_dense_impl(const VectorType* b, VectorType* x,
+                                 initial_guess_mode guess) const
+{
+    using ws = workspace_traits<Multigrid>;
+    this->setup_workspace();
+    this->create_state();
+    if (cache_.state->nrhs != b->get_size()[1]) {
+        cache_.state->generate(this->get_system_matrix().get(), this,
+                               b->get_size()[1]);
+    }
+    auto lambda = [&, this](auto mg_level, auto b, auto x) {
         using value_type = typename std::decay_t<
             gko::detail::pointee<decltype(mg_level)>>::value_type;
         auto exec = this->get_executor();
-        auto neg_one_op =
+        static auto neg_one_op =
             initialize<matrix::Dense<value_type>>({-one<value_type>()}, exec);
-        auto one_op =
+        static auto one_op =
             initialize<matrix::Dense<value_type>>({one<value_type>()}, exec);
```

Comment:
Why don't you put them into the `cache_` as well? A `static` means you have an atomic check each time these lines are reached.

Reply:
I take these two from the `cache_.state` now